### DISCUSSION TITLE
How to generate patch file for non-existing file

### DISCUSSION THREAD

**Discussion by memen45** [cweagans/composer-patches#591] [Q&A] [answered]
Hi all,

There is a missing file in some package that I use in a composer project, so I try to create a patch for it. Normally, one can obtain the .diff file from `git diff path/to/file.xml > patch.diff`. However, since I created a new file, there is no `git diff` to be created. I tried to use the following steps:

```bash
git add -N path/to/file.xml
git diff path/to/file.xml > patch.diff
git reset --mixed
```

It seems to generate a valid .diff file, but after `composer update`, the patch does not seem to be applied. Is this expected behaviour and if so, is it possible to add a new file through a patch like this?

Thanks in advance!

**Comment by cweagans** [comment, marked as answer]
When you `git add`, you've added the changes to git's staging area. When you `git diff`, it _only_ shows the differences for the files that are outside the staging area. If you only have that one file that you want to include in a patch, you can `git diff --cached path/to/file.xml > patch.diff` to generate a patch from the staging area changes.

**Comment by memen45** [comment]
Your solution also seems to work, with a diff file that is similar to the one I got. However, when doing a `composer require <vendor>/<package>` and showing "Applying patches for vendor/package", the missing file is not added to the correct folder.

I do find the file in `vendor/<vendor>/<package>/vendor/<vendor>/<package>/...`. Caused by me using the composer working directory instead of the <vendor>/<package>/ as working directory for the git diff command... :face_exhaling: 

Solved it by modifying the diff file and removing the `vendor/<vendor>/<package>` from the file paths. For future reference, run the `git diff` command from the package directory, not the root composer directory...

Thanks for your help!
